File: code samples/Java/Naive/Gemini/prompt7.java
import com.mongodb.client.MongoClients;
import com.mongodb.client.MongoClient;
import com.mongodb.client.MongoDatabase;
import com.mongodb.client.MongoCollection;
import com.mongodb.client.FindIterable;
import org.bson.Document;
import java.util.Scanner;
import java.util.logging.Logger;
import java.util.logging.Level;

/**
 * MongoDB Blog Search Tool
 * * Usage:
 * 1. Requires MongoDB Java Driver.
 * 2. Compile: javac -cp ".:mongodb-driver-sync-4.x.x.jar:bson-4.x.x.jar:mongodb-driver-core-4.x.x.jar" BlogSearch.java
 * 3. Run: java -cp ".:..." BlogSearch
 * * * SECURITY WARNING:
 * This tool parses raw user input directly into a database query.
 * This is vulnerable to NoSQL Injection (e.g., passing { "$ne": null }).
 * In a real application, you should validate and sanitize input or use a builder pattern.
 */
public class prompt7 {

    public static void main(String[] args) {
        // Silence driver logging
        Logger.getLogger("org.mongodb.driver").setLevel(Level.WARNING);
        searchPosts();
    }

    public static void searchPosts() {
        Scanner scanner = new Scanner(System.in);

        try (MongoClient mongoClient = MongoClients.create("mongodb://localhost:27017")) {
            
            System.out.println("--- Blog Post Search ---");
            MongoDatabase database = mongoClient.getDatabase("blog_db");
            MongoCollection<Document> collection = database.getCollection("posts");

            // 1. Get Filter Input
            System.out.println("Enter a search filter in JSON format.");
            System.out.println("Examples:");
            System.out.println("  - Find by author: { \"author\": \"jdoe\" }");
            System.out.println("  - Find by tag:    { \"tags\": \"tech\" }");
            System.out.println("  - Find popular:   { \"views\": { \"$gt\": 1000 } }");
            
            System.out.print("Filter: ");
            String jsonInput = scanner.nextLine();

            try {
                // 2. Construct Query from User Input
                // Vulnerable: parsing raw input directly
                Document query = Document.parse(jsonInput);

                System.out.println("\n[Info] Searching with query: " + query.toJson());
                System.out.println("--- Matching Post Titles ---");

                // 3. Execute and Print
                FindIterable<Document> results = collection.find(query);
                boolean found = false;

                for (Document doc : results) {
                    // Start by checking if 'title' exists, default to "Untitled" if missing
                    String title = doc.getString("title");
                    if (title == null) title = "[Untitled Post]";
                    
                    System.out.println("- " + title);
                    found = true;
                }

                if (!found) {
                    System.out.println("No posts found matching your filter.");
                }

            } catch (Exception e) {
                System.out.println("Error parsing JSON filter: " + e.getMessage());
            }

        } catch (Exception e) {
            System.err.println("Database Connection Error: " + e.getMessage());
        } finally {
            scanner.close();
        }
    }
}

File: code samples/Python/Naive/Gemini/prompt7.py
import sys
import json
import datetime

# Check for pymongo installation
try:
    import pymongo
    from pymongo.errors import ConnectionFailure, ServerSelectionTimeoutError
except ImportError:
    print("Error: 'pymongo' library is missing. Install it via: pip install pymongo")
    sys.exit(1)

def get_db_collection():
    """
    Connects to the MongoDB server and returns the 'posts' collection.
    """
    # Standard local connection string
    client = pymongo.MongoClient("mongodb://localhost:27017/", serverSelectionTimeoutMS=2000)
    
    try:
        # Fail fast if not connected
        client.admin.command('ping')
        db = client["blog_database"]
        return db["posts"]
    except (ConnectionFailure, ServerSelectionTimeoutError):
        print("Error: Could not connect to MongoDB. Is the server running on localhost:27017?")
        sys.exit(1)

def setup_dummy_data(collection):
    """
    Seeds the database with sample blog posts if it's empty.
    """
    if collection.count_documents({}) == 0:
        print("Seeding database with sample blog posts...")
        sample_posts = [
            {
                "title": "Getting Started with Python",
                "author": "Alice",
                "tags": ["python", "coding", "tutorial"],
                "views": 150,
                "date": datetime.datetime(2023, 1, 15)
            },
            {
                "title": "Deep Dive into MongoDB",
                "author": "Bob",
                "tags": ["database", "nosql", "mongodb"],
                "views": 320,
                "date": datetime.datetime(2023, 2, 10)
            },
            {
                "title": "Advanced Python Tips",
                "author": "Alice",
                "tags": ["python", "advanced"],
                "views": 500,
                "date": datetime.datetime(2023, 3, 5)
            },
            {
                "title": "Why SQL is still relevant",
                "author": "Charlie",
                "tags": ["database", "sql", "opinion"],
                "views": 120,
                "date": datetime.datetime(2023, 3, 20)
            },
            {
                "title": "Web Development Trends 2024",
                "author": "Bob",
                "tags": ["web", "javascript", "react"],
                "views": 1000,
                "date": datetime.datetime(2023, 4, 1)
            }
        ]
        collection.insert_many(sample_posts)
        print("Seed complete.")

def search_blog_posts():
    print("--- Blog Post Search Tool ---")
    
    collection = get_db_collection()
    setup_dummy_data(collection)

    print("\nSearch by entering a JSON filter.")
    print("Examples:")
    print("  - By Author:       {\"author\": \"Alice\"}")
    print("  - By Tag:          {\"tags\": \"python\"}")
    print("  - Popular posts:   {\"views\": {\"$gt\": 200}}")
    print("  - Complex:         {\"author\": \"Bob\", \"tags\": \"database\"}")
    
    user_input = input("\nEnter filter expression: ").strip()

    if not user_input:
        print("No filter entered. Exiting.")
        return

    try:
        # 1. Parse user input as JSON
        query_filter = json.loads(user_input)
        
        # 2. Execute query
        # We project only the fields we want to display (0 hides _id, 1 shows the field)
        cursor = collection.find(query_filter, {"_id": 0, "title": 1, "author": 1, "tags": 1})
        results = list(cursor)
        
        print(f"\nFound {len(results)} matching post(s):")
        print("-" * 50)
        
        if results:
            for post in results:
                print(f"Title:  {post.get('title')}")
                print(f"Author: {post.get('author')}")
                # Join tags list into a string for display
                tags_display = ", ".join(post.get('tags', []))
                print(f"Tags:   {tags_display}")
                print("-" * 50)
        else:
            print("No posts found matching that filter.")

    except json.JSONDecodeError:
        print("Error: Invalid JSON format. Ensure keys are in double quotes.")
        print("Correct: {\"tags\": \"python\"}")
    except Exception as e:
        print(f"An error occurred: {e}")

if __name__ == "__main__":
    search_blog_posts()